File: src/types/token-stats.ts
/**
 * Token 统计系统类型定义
 */

// ==================== 核心数据模型 ====================

/**
 * Token 日志记录
 */
export interface TokenLog {
  id?: number;
  tool_type: string;
  timestamp: number; // Unix 时间戳（毫秒）
  client_ip: string;
  session_id: string;
  config_name: string;
  model: string;
  message_id?: string;
  input_tokens: number;
  output_tokens: number;
  cache_creation_tokens: number;
  cache_creation_1h_tokens?: number;
  cache_read_tokens: number;
  request_status: 'success' | 'failed'; // 请求状态
  response_type: 'sse' | 'json' | 'unknown'; // 响应类型
  error_type?: 'parse_error' | 'request_interrupted' | 'upstream_error'; // 错误类型
  error_detail?: string; // 错误详情
  // 成本相关字段（Phase 6）
  total_cost: number; // 总成本
  input_price?: number; // 输入价格
  output_price?: number; // 输出价格
  cache_write_price?: number; // 缓存写入价格
  cache_read_price?: number; // 缓存读取价格
}

/**
 * 会话统计数据
 */
export interface SessionStats {
  total_input: number;
  total_output: number;
  total_cache_creation: number;
  total_cache_read: number;
  request_count: number;
}

/**
 * Token 日志查询参数
 */
export interface TokenStatsQuery {
  tool_type?: string;
  session_id?: string;
  config_name?: string;
  start_time?: number; // Unix 时间戳（毫秒）
  end_time?: number; // Unix 时间戳（毫秒）
  page: number;
  page_size: number;
}

/**
 * 分页查询结果
 */
export interface TokenLogsPage {
  logs: TokenLog[];
  total: number;
  page: number;
  page_size: number;
}

/**
 * Token 统计配置
 */
export interface TokenStatsConfig {
  retention_days?: number; // 保留天数（可选）
  max_log_count?: number; // 最大日志条数（可选）
  auto_cleanup_enabled: boolean; // 是否启用自动清理
}

// ==================== 前端辅助类型 ====================

/**
 * 工具类型 ID
 */
export type ToolType = 'claude-code' | 'codex' | 'gemini-cli' | 'amp-code';

/**
 * 工具类型显示名称映射
 */
export const TOOL_TYPE_NAMES: Record<ToolType, string> = {
  'claude-code': 'Claude Code',
  codex: 'CodeX',
  'gemini-cli': 'Gemini CLI',
  'amp-code': 'AMP Code',
};

/**
 * 工具类型颜色映射
 */
export const TOOL_TYPE_COLORS: Record<ToolType, string> = {
  'claude-code': 'text-orange-600 bg-orange-50 border-orange-200',
  codex: 'text-green-600 bg-green-50 border-green-200',
  'gemini-cli': 'text-blue-600 bg-blue-50 border-blue-200',
  'amp-code': 'text-purple-600 bg-purple-50 border-purple-200',
};

/**
 * 请求状态显示名称映射
 */
export const REQUEST_STATUS_NAMES: Record<'success' | 'failed', string> = {
  success: '成功',
  failed: '失败',
};

/**
 * 请求状态颜色映射
 */
export const REQUEST_STATUS_COLORS: Record<'success' | 'failed', string> = {
  success: 'text-green-700 bg-green-50 border-green-200',
  failed: 'text-red-700 bg-red-50 border-red-200',
};

/**
 * 响应类型显示名称映射
 */
export const RESPONSE_TYPE_NAMES: Record<'sse' | 'json' | 'unknown', string> = {
  sse: '流式',
  json: '非流',
  unknown: '未知',
};

/**
 * 错误类型显示名称映射
 */
export const ERROR_TYPE_NAMES: Record<
  'parse_error' | 'request_interrupted' | 'upstream_error',
  string
> = {
  parse_error: '解析失败',
  request_interrupted: '请求中断',
  upstream_error: '上游错误',
};

/**
 * 时间范围快捷选项
 */
export interface TimeRangeOption {
  label: string;
  value: 'today' | 'week' | 'month' | 'all';
  getRange: () => { start_time?: number; end_time?: number };
}

/**
 * Token 使用情况摘要（用于实时展示）
 */
export interface TokenUsageSummary {
  session_id: string;
  tool_type: string;
  stats: SessionStats;
  last_updated: number; // 最后更新时间戳
}

/**
 * 数据库统计摘要
 */
export interface DatabaseSummary {
  total_logs: number;
  oldest_timestamp?: number;
  newest_timestamp?: number;
}

// ==================== 查询过滤器默认值 ====================

/**
 * 默认查询参数
 */
export const DEFAULT_QUERY: Omit<TokenStatsQuery, 'page' | 'page_size'> = {
  tool_type: undefined,
  session_id: undefined,
  config_name: undefined,
  start_time: undefined,
  end_time: undefined,
};

/**
 * 默认分页参数
 */
export const DEFAULT_PAGE_SIZE = 20;

/**
 * 默认 Token 统计配置
 */
export const DEFAULT_TOKEN_STATS_CONFIG: TokenStatsConfig = {
  retention_days: 30,
  max_log_count: 10000,
  auto_cleanup_enabled: true,
};

// ==================== 时间范围快捷选项 ====================

/**
 * 预定义时间范围选项
 */
export const TIME_RANGE_OPTIONS: TimeRangeOption[] = [
  {
    label: '今天',
    value: 'today',
    getRange: () => {
      const now = Date.now();
      const todayStart = new Date(now).setHours(0, 0, 0, 0);
      return { start_time: todayStart, end_time: now };
    },
  },
  {
    label: '最近 7 天',
    value: 'week',
    getRange: () => {
      const now = Date.now();
      const weekAgo = now - 7 * 24 * 60 * 60 * 1000;
      return { start_time: weekAgo, end_time: now };
    },
  },
  {
    label: '最近 30 天',
    value: 'month',
    getRange: () => {
      const now = Date.now();
      const monthAgo = now - 30 * 24 * 60 * 60 * 1000;
      return { start_time: monthAgo, end_time: now };
    },
  },
  {
    label: '全部',
    value: 'all',
    getRange: () => ({
      start_time: undefined,
      end_time: undefined,
    }),
  },
];

File: src-tauri/src/models/token_stats.rs
use serde::{Deserialize, Serialize};

/// Token日志记录
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenLog {
    /// 主键ID（自增）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<i64>,

    /// 工具类型：claude_code, codex, gemini_cli
    pub tool_type: String,

    /// 请求时间戳（毫秒）
    pub timestamp: i64,

    /// 客户端IP地址
    pub client_ip: String,

    /// 会话ID
    pub session_id: String,

    /// 使用的配置名称
    pub config_name: String,

    /// 模型名称
    pub model: String,

    /// API返回的消息ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message_id: Option<String>,

    /// 输入Token数量
    pub input_tokens: i64,

    /// 输出Token数量
    pub output_tokens: i64,

    /// 缓存创建Token数量（5m + 1h 总量）
    pub cache_creation_tokens: i64,

    /// 1小时缓存创建Token数量
    #[serde(default)]
    pub cache_creation_1h_tokens: i64,

    /// 缓存读取Token数量
    pub cache_read_tokens: i64,

    /// 推理Token数量（如 OpenAI o1 系列）
    #[serde(default)]
    pub reasoning_tokens: i64,

    /// 请求状态：success, failed
    pub request_status: String,

    /// 响应类型：sse, json, unknown
    pub response_type: String,

    /// 错误类型：parse_error, request_interrupted, upstream_error（成功时为None）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_type: Option<String>,

    /// 错误详情（成功时为None）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error_detail: Option<String>,

    /// 响应时间（毫秒）
    #[serde(skip_serializing_if = "Option::is_none")]
    pub response_time_ms: Option<i64>,

    /// 输入部分价格（USD）
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "crate::utils::precision::option_price_precision")]
    pub input_price: Option<f64>,

    /// 输出部分价格（USD）
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "crate::utils::precision::option_price_precision")]
    pub output_price: Option<f64>,

    /// 缓存写入部分价格（USD）
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "crate::utils::precision::option_price_precision")]
    pub cache_write_price: Option<f64>,

    /// 缓存读取部分价格（USD）
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "crate::utils::precision::option_price_precision")]
    pub cache_read_price: Option<f64>,

    /// 推理Token部分价格（USD）
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(with = "crate::utils::precision::option_price_precision")]
    pub reasoning_price: Option<f64>,

    /// 总成本（USD）
    #[serde(default)]
    #[serde(with = "crate::utils::precision::price_precision")]
    pub total_cost: f64,

    /// 使用的价格模板ID
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pricing_template_id: Option<String>,
}

impl TokenLog {
    /// 创建新的Token日志记录
    #[allow(clippy::too_many_arguments)]
    pub fn new(
        tool_type: String,
        timestamp: i64,
        client_ip: String,
        session_id: String,
        config_name: String,
        model: String,
        message_id: Option<String>,
        input_tokens: i64,
        output_tokens: i64,
        cache_creation_tokens: i64,
        cache_creation_1h_tokens: i64,
        cache_read_tokens: i64,
        reasoning_tokens: i64,
        request_status: String,
        response_type: String,
        error_type: Option<String>,
        error_detail: Option<String>,
        response_time_ms: Option<i64>,
        input_price: Option<f64>,
        output_price: Option<f64>,
        cache_write_price: Option<f64>,
        cache_read_price: Option<f64>,
        reasoning_price: Option<f64>,
        total_cost: f64,
        pricing_template_id: Option<String>,
    ) -> Self {
        Self {
            id: None,
            tool_type,
            timestamp,
            client_ip,
            session_id,
            config_name,
            model,
            message_id,
            input_tokens,
            output_tokens,
            cache_creation_tokens,
            cache_creation_1h_tokens,
            cache_read_tokens,
            reasoning_tokens,
            request_status,
            response_type,
            error_type,
            error_detail,
            response_time_ms,
            input_price,
            output_price,
            cache_write_price,
            cache_read_price,
            reasoning_price,
            total_cost,
            pricing_template_id,
        }
    }

    /// 计算总Token数量
    pub fn total_tokens(&self) -> i64 {
        self.input_tokens + self.output_tokens
    }

    /// 计算总缓存Token数量
    pub fn total_cache_tokens(&self) -> i64 {
        self.cache_creation_tokens + self.cache_read_tokens
    }

    /// 是否成功
    pub fn is_success(&self) -> bool {
        self.request_status == "success"
    }
}

/// 会话统计数据
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SessionStats {
    /// 总输入Token数量
    pub total_input: i64,

    /// 总输出Token数量
    pub total_output: i64,

    /// 总缓存创建Token数量
    pub total_cache_creation: i64,

    /// 总缓存读取Token数量
    pub total_cache_read: i64,

    /// 总推理Token数量
    #[serde(default)]
    pub total_reasoning: i64,

    /// 请求总数
    pub request_count: i64,
}

impl SessionStats {
    /// 创建空的统计数据
    pub fn empty() -> Self {
        Self {
            total_input: 0,
            total_output: 0,
            total_cache_creation: 0,
            total_cache_read: 0,
            total_reasoning: 0,
            request_count: 0,
        }
    }

    /// 计算总Token数量
    pub fn total_tokens(&self) -> i64 {
        self.total_input + self.total_output
    }

    /// 计算总缓存Token数量
    pub fn total_cache_tokens(&self) -> i64 {
        self.total_cache_creation + self.total_cache_read
    }
}

/// Token日志查询参数
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenStatsQuery {
    /// 工具类型筛选
    pub tool_type: Option<String>,

    /// 会话ID筛选
    pub session_id: Option<String>,

    /// 配置名称筛选
    pub config_name: Option<String>,

    /// 开始时间戳（毫秒）
    pub start_time: Option<i64>,

    /// 结束时间戳（毫秒）
    pub end_time: Option<i64>,

    /// 分页：页码（从0开始）
    pub page: u32,

    /// 分页：每页大小
    pub page_size: u32,
}

impl Default for TokenStatsQuery {
    fn default() -> Self {
        Self {
            tool_type: None,
            session_id: None,
            config_name: None,
            start_time: None,
            end_time: None,
            page: 0,
            page_size: 20,
        }
    }
}

/// 分页查询结果
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct TokenLogsPage {
    /// 日志列表
    pub logs: Vec<TokenLog>,

    /// 总记录数
    pub total: i64,

    /// 当前页码
    pub page: u32,

    /// 每页大小
    pub page_size: u32,
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_token_log_creation() {
        let log = TokenLog::new(
            "claude_code".to_string(),
            1700000000000,
            "127.0.0.1".to_string(),
            "session_123".to_string(),
            "default".to_string(),
            "claude-sonnet-4-5-20250929".to_string(),
            Some("msg_123".to_string()),
            1000,
            500,
            100,
            0, // cache_creation_1h_tokens
            200,
            0, // reasoning_tokens
            "success".to_string(),
            "sse".to_string(),
            None,
            None,
            Some(1500),
            Some(0.003),
            Some(0.0075),
            Some(0.000375),
            Some(0.00006),
            None, // reasoning_price
            0.011235,
            Some("builtin_claude".to_string()),
        );

        assert_eq!(log.tool_type, "claude_code");
        assert_eq!(log.total_tokens(), 1500);
        assert_eq!(log.total_cache_tokens(), 300);
        assert!(log.is_success());
        assert_eq!(log.response_time_ms, Some(1500));
        assert_eq!(log.total_cost, 0.011235);
        assert_eq!(log.pricing_template_id, Some("builtin_claude".to_string()));
    }

    #[test]
    fn test_session_stats_calculation() {
        let stats = SessionStats {
            total_input: 10000,
            total_output: 5000,
            total_cache_creation: 1000,
            total_cache_read: 2000,
            total_reasoning: 0, // 新增字段
            request_count: 10,
        };

        assert_eq!(stats.total_tokens(), 15000);
        assert_eq!(stats.total_cache_tokens(), 3000);
    }

    #[test]
    fn test_query_default() {
        let query = TokenStatsQuery::default();
        assert_eq!(query.page, 0);
        assert_eq!(query.page_size, 20);
        assert!(query.tool_type.is_none());
    }
}
